File: src/ios/FirebasexPerformancePlugin.m
/**
 * @file FirebasexPerformancePlugin.m
 * @brief iOS implementation of the Firebase Performance Monitoring Cordova plugin.
 */

#import "FirebasexPerformancePlugin.h"
#import "FirebasexCorePlugin.h"
@import FirebasePerformance;

/** Preference key for the performance data collection enabled state. */
static NSString *const FIREBASE_PERFORMANCE_COLLECTION_ENABLED =
    @"FIREBASE_PERFORMANCE_COLLECTION_ENABLED";

@implementation FirebasexPerformancePlugin {
    /** Dictionary of active traces, keyed by trace name. Thread-safe via @synchronized. */
    NSMutableDictionary *_traces;
}

/**
 * Initialises the plugin, creating the traces dictionary and reading the
 * persisted performance collection preference.
 */
- (void)pluginInitialize {
    NSLog(@"FirebasexPerformancePlugin pluginInitialize");
    _traces = [[NSMutableDictionary alloc] init];

    FirebasexCorePlugin *core = [FirebasexCorePlugin sharedInstance];
    if ([core getPreferenceFlag:FIREBASE_PERFORMANCE_COLLECTION_ENABLED]) {
        [core setPreferenceFlag:FIREBASE_PERFORMANCE_COLLECTION_ENABLED flag:YES];
    }
}

/**
 * Enables or disables Firebase Performance data collection.
 *
 * Persists the setting via the core plugin's preference flags.
 *
 * @param command Cordova command. args[0] = boolean.
 */
- (void)setPerformanceCollectionEnabled:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        @try {
            BOOL enabled = [[command argumentAtIndex:0] boolValue];
            [[FIRPerformance sharedInstance] setDataCollectionEnabled:enabled];
            [[FirebasexCorePlugin sharedInstance] setPreferenceFlag:FIREBASE_PERFORMANCE_COLLECTION_ENABLED
                                                              flag:enabled];
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } @catch (NSException *exception) {
            [[FirebasexCorePlugin sharedInstance] handlePluginExceptionWithContext:exception :command];
        }
    }];
}

/**
 * Returns the current performance data collection enabled state as a boolean.
 */
- (void)isPerformanceCollectionEnabled:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        @try {
            CDVPluginResult *pluginResult = [CDVPluginResult
                resultWithStatus:CDVCommandStatus_OK
                   messageAsBool:[[FirebasexCorePlugin sharedInstance]
                                     getPreferenceFlag:FIREBASE_PERFORMANCE_COLLECTION_ENABLED]];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } @catch (NSException *exception) {
            [[FirebasexCorePlugin sharedInstance] handlePluginExceptionWithContext:exception :command];
        }
    }];
}

/**
 * Starts a custom performance trace.
 *
 * If a trace with the given name already exists, it is reused.
 * The trace dictionary is protected with @c @synchronized for thread safety.
 *
 * @param command Cordova command. args[0] = trace name (NSString).
 */
- (void)startTrace:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        @try {
            NSString *traceName = [command.arguments objectAtIndex:0];
            @synchronized(self->_traces) {
                FIRTrace *trace = [self->_traces objectForKey:traceName];
                if (trace == nil) {
                    trace = [FIRPerformance startTraceWithName:traceName];
                    [self->_traces setObject:trace forKey:traceName];
                }
            }
            CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } @catch (NSException *exception) {
            [[FirebasexCorePlugin sharedInstance] handlePluginExceptionWithContext:exception :command];
        }
    }];
}

/**
 * Increments a named metric on an active trace by 1.
 *
 * Returns an error if the trace is not found.
 *
 * @param command Cordova command. args[0] = trace name, args[1] = counter name.
 */
- (void)incrementCounter:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        @try {
            NSString *traceName = [command.arguments objectAtIndex:0];
            NSString *counterNamed = [command.arguments objectAtIndex:1];
            CDVPluginResult *pluginResult;
            @synchronized(self->_traces) {
                FIRTrace *trace = (FIRTrace *)[self->_traces objectForKey:traceName];
                if (trace != nil) {
                    [trace incrementMetric:counterNamed byInt:1];
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                } else {
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                     messageAsString:@"Trace not found"];
                }
            }
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } @catch (NSException *exception) {
            [[FirebasexCorePlugin sharedInstance] handlePluginExceptionWithContext:exception :command];
        }
    }];
}

/**
 * Stops an active trace, submits its data, and removes it from the dictionary.
 *
 * Returns an error if the trace is not found.
 *
 * @param command Cordova command. args[0] = trace name.
 */
- (void)stopTrace:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        @try {
            NSString *traceName = [command.arguments objectAtIndex:0];
            CDVPluginResult *pluginResult;
            @synchronized(self->_traces) {
                FIRTrace *trace = [self->_traces objectForKey:traceName];
                if (trace != nil) {
                    [trace stop];
                    [self->_traces removeObjectForKey:traceName];
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
                } else {
                    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                     messageAsString:@"Trace not found"];
                }
            }
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        } @catch (NSException *exception) {
            [[FirebasexCorePlugin sharedInstance] handlePluginExceptionWithContext:exception :command];
        }
    }];
}

@end

File: src/android/FirebasexPerformancePlugin.java
package org.apache.cordova.firebasex;

import android.util.Log;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;

import com.google.firebase.perf.FirebasePerformance;
import com.google.firebase.perf.metrics.Trace;

import java.util.HashMap;

/**
 * Cordova plugin for Firebase Performance Monitoring on Android.
 *
 * <p>Provides custom trace management (start, increment counter, stop) and
 * control over performance data collection. Active traces are stored in an
 * in-memory map keyed by trace name.
 *
 * <p>The performance collection enabled state is persisted via
 * {@link FirebasexCorePlugin} shared preferences so it survives app restarts.
 *
 * @see <a href="https://firebase.google.com/docs/perf-mon">Firebase Performance Monitoring</a>
 */
public class FirebasexPerformancePlugin extends CordovaPlugin {

    /** Log tag for all messages from this plugin. */
    private static final String TAG = "FirebasexPerformance";

    /** SharedPreferences key for the performance collection enabled state. */
    private static final String PERFORMANCE_COLLECTION_ENABLED = "firebase_performance_collection_enabled";

    /** Map of active traces, keyed by trace name. */
    private HashMap<String, Trace> traces = new HashMap<String, Trace>();

    /**
     * Initialises the plugin.
     *
     * Checks the shared preference for performance collection enabled state.
     * If not already set, reads the value from AndroidManifest meta-data and
     * persists it to shared preferences.
     */
    @Override
    protected void pluginInitialize() {
        Log.d(TAG, "pluginInitialize");
        try {
            boolean enabled = FirebasexCorePlugin.getInstance().getPreference(PERFORMANCE_COLLECTION_ENABLED);
            if (!enabled) {
                android.os.Bundle metaData = cordova.getActivity().getPackageManager()
                        .getApplicationInfo(cordova.getActivity().getPackageName(),
                                android.content.pm.PackageManager.GET_META_DATA).metaData;
                if (metaData != null && metaData.containsKey(PERFORMANCE_COLLECTION_ENABLED)) {
                    boolean metaEnabled = metaData.getBoolean(PERFORMANCE_COLLECTION_ENABLED);
                    if (metaEnabled) {
                        FirebasexCorePlugin.getInstance().setPreference(PERFORMANCE_COLLECTION_ENABLED, true);
                    }
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error reading performance collection preference: " + e.getMessage());
        }
    }

    /**
     * Dispatches Cordova actions to plugin methods.
     *
     * <p>Supported actions:
     * <ul>
     *   <li>{@code "startTrace"} - start a custom trace</li>
     *   <li>{@code "incrementCounter"} - increment a metric on an active trace</li>
     *   <li>{@code "stopTrace"} - stop and submit an active trace</li>
     *   <li>{@code "setPerformanceCollectionEnabled"} - enable/disable data collection</li>
     *   <li>{@code "isPerformanceCollectionEnabled"} - query data collection state</li>
     * </ul>
     */
    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        switch (action) {
            case "startTrace":
                this.startTrace(callbackContext, args.getString(0));
                return true;
            case "incrementCounter":
                this.incrementCounter(callbackContext, args.getString(0), args.getString(1));
                return true;
            case "stopTrace":
                this.stopTrace(callbackContext, args.getString(0));
                return true;
            case "setPerformanceCollectionEnabled":
                this.setPerformanceCollectionEnabled(callbackContext, args.getBoolean(0));
                return true;
            case "isPerformanceCollectionEnabled":
                this.isPerformanceCollectionEnabled(callbackContext);
                return true;
        }
        return false;
    }

    /**
     * Starts a custom performance trace with the given name.
     *
     * If a trace with the same name already exists, the existing trace is reused.
     *
     * @param callbackContext the Cordova callback
     * @param name            the trace name
     */
    private void startTrace(final CallbackContext callbackContext, final String name) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    Trace myTrace = traces.get(name);
                    if (myTrace == null) {
                        myTrace = FirebasePerformance.getInstance().newTrace(name);
                        myTrace.start();
                        traces.put(name, myTrace);
                    }
                    callbackContext.success();
                } catch (Exception e) {
                    FirebasexCorePlugin.handleExceptionWithContext(e, callbackContext);
                }
            }
        });
    }

    /**
     * Increments a named metric on an active trace by 1.
     *
     * @param callbackContext the Cordova callback
     * @param name            the trace name
     * @param counterNamed    the metric/counter name to increment
     */
    private void incrementCounter(final CallbackContext callbackContext, final String name, final String counterNamed) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    Trace myTrace = traces.get(name);
                    if (myTrace != null) {
                        myTrace.incrementMetric(counterNamed, 1);
                        callbackContext.success();
                    } else {
                        callbackContext.error("Trace not found");
                    }
                } catch (Exception e) {
                    FirebasexCorePlugin.handleExceptionWithContext(e, callbackContext);
                }
            }
        });
    }

    /**
     * Stops an active trace, submitting its data, and removes it from the map.
     *
     * @param callbackContext the Cordova callback
     * @param name            the trace name to stop
     */
    private void stopTrace(final CallbackContext callbackContext, final String name) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    Trace myTrace = traces.get(name);
                    if (myTrace != null) {
                        myTrace.stop();
                        traces.remove(name);
                        callbackContext.success();
                    } else {
                        callbackContext.error("Trace not found");
                    }
                } catch (Exception e) {
                    FirebasexCorePlugin.handleExceptionWithContext(e, callbackContext);
                }
            }
        });
    }

    /**
     * Enables or disables Firebase Performance Monitoring data collection.
     *
     * The setting is persisted to shared preferences via the core plugin.
     *
     * @param callbackContext the Cordova callback
     * @param enabled         {@code true} to enable, {@code false} to disable
     */
    private void setPerformanceCollectionEnabled(final CallbackContext callbackContext, final boolean enabled) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    FirebasePerformance.getInstance().setPerformanceCollectionEnabled(enabled);
                    FirebasexCorePlugin.getInstance().setPreference(PERFORMANCE_COLLECTION_ENABLED, enabled);
                    callbackContext.success();
                } catch (Exception e) {
                    FirebasexCorePlugin.handleExceptionWithContext(e, callbackContext);
                }
            }
        });
    }

    /**
     * Queries the current performance data collection enabled state from shared preferences.
     *
     * Returns 1 (enabled) or 0 (disabled) to the callback.
     *
     * @param callbackContext the Cordova callback
     */
    private void isPerformanceCollectionEnabled(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                try {
                    boolean enabled = FirebasexCorePlugin.getInstance().getPreference(PERFORMANCE_COLLECTION_ENABLED);
                    callbackContext.success(enabled ? 1 : 0);
                } catch (Exception e) {
                    FirebasexCorePlugin.handleExceptionWithContext(e, callbackContext);
                }
            }
        });
    }
}
